river/Layout.zig

```diff
@@ -0,0 +1,182 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const std = @import("std");
+const wlr = @import("wlroots");
+const wayland = @import("wayland");
+const wl = wayland.server.wl;
+const zriver = wayland.server.zriver;
+
+const log = @import("log.zig");
+const util = @import("util.zig");
+
+const Box = @import("Box.zig");
+const Server = @import("Server.zig");
+const Output = @import("Output.zig");
+const View = @import("View.zig");
+const ViewStack = @import("view_stack.zig").ViewStack;
+const LayoutDemand = @import("LayoutDemand.zig");
+
+layout: *zriver.LayoutV1,
+namespace: ?[]const u8,
+output: *Output,
+
+pub fn init(self: *Self, namespace: [*:0]const u8, output: *Output, client: *wl.Client, version: u32, id: u32) !void {
+    self.output = output;
+    self.namespace = try util.gpa.dupe(u8, std.mem.span(namespace));
+    self.layout = try zriver.LayoutV1.create(client, version, id);
+    self.layout.setHandler(*Self, handleRequest, handleDestroy, self);
+
+    var output_it = output.root.outputs.first;
+    while (output_it) |output_node| : (output_it = output_node.next) {
+        var layout_it = output_node.data.layouts.first;
+        if (output_node.data.wlr_output == output.wlr_output) {
+
+            // On this output, no other layout can have our namespace.
+            while (layout_it) |layout_node| : (layout_it = layout_node.next) {
+                if (std.mem.eql(u8, self.namespace.?, layout_node.data.namespace.?)) {
+                    self.layout.sendNamespaceInUse();
+                    util.gpa.free(self.namespace.?);
+                    self.namespace = null;
+                    return;
+                }
+            }
+        } else {
+
+            // Layouts on other outputs may share the namespace, if they come from the same client.
+            while (layout_it) |layout_node| : (layout_it = layout_node.next) {
+                if (std.mem.eql(u8, self.namespace.?, layout_node.data.namespace.?) and
+                    layout_node.data.layout.getClient() != self.layout.getClient())
+                {
+                    self.layout.sendNamespaceInUse();
+                    util.gpa.free(self.namespace.?);
+                    self.namespace = null;
+                    return;
+                }
+            }
+        }
+    }
+}
+
+/// Send a layout demand to the client
+pub fn startLayoutDemand(self: *Self, views: u32) void {
+    log.debug(
+        .layout_manager,
+        "starting layout demand '{}' on output '{}'",
+        .{ self.namespace, self.output.wlr_output.name },
+    );
+
+    // First we create the Layoutdemand object
+    var demand: LayoutDemand = .{};
+    demand.init(self.output, views);
+    self.output.layout_demand = demand;
```

Comment:
We don't store pointers to the layout demand anywhere, so we can use the more idiomatic
```suggestion
    self.output.layout_demand = LayoutDemand.init(self.output, views);
```